services/github/github-issue-detail.service.js

```diff
@@ -1,27 +1,142 @@
 'use strict'
 
-const LegacyService = require('../legacy-service')
-const {
-  makeLabel: getLabel,
-  makeBadgeData: getBadgeData,
-} = require('../../lib/badge-data')
-const { makeLogo: getLogo } = require('../../lib/logos')
-const { formatDate } = require('../text-formatters')
-const { age: ageColor } = require('../color-formatters')
+const Joi = require('joi')
+const { InvalidResponse } = require('..')
+const { nonNegativeInteger } = require('../validators')
+const { formatDate, metric } = require('../text-formatters')
+const { age } = require('../color-formatters')
+const { GithubAuthService } = require('./github-auth-service')
 const {
   documentation,
-  stateColor: githubStateColor,
-  commentsColor: githubCommentsColor,
-  checkErrorResponse: githubCheckErrorResponse,
+  errorMessagesFor,
+  stateColor,
+  commentsColor,
 } = require('./github-helpers')
 
-// This legacy service should be rewritten to use e.g. BaseJsonService.
-//
-// Tips for rewriting:
-// https://github.com/badges/shields/blob/master/doc/rewriting-services.md
-//
-// Do not base new services on this code.
-module.exports = class GithubIssueDetail extends LegacyService {
+const getDefaultLabel = ({ json }) => {
+  if (json.pull_request) {
+    return `pull request ${json.number}`
+  }
+
+  return `issue ${json.number}`
+}
+
+const commonSchemaFields = {
+  number: nonNegativeInteger,
+  pull_request: Joi.any(),
+}
+
+const stateMap = {
+  schema: Joi.object({
+    ...commonSchemaFields,
+    state: Joi.string()
+      .allow('open', 'closed')
+      .required(),
+  }).required(),
+  transform: ({ json }) => json.state,
+  render: ({ value, json }) => ({
+    color: stateColor(value),
+    label: getDefaultLabel({ json }),
+    message: value,
+  }),
+}
+
+const titleMap = {
+  schema: Joi.object({
+    ...commonSchemaFields,
+    title: Joi.string().required(),
+  }).required(),
+  transform: ({ json }) => json.title,
+  render: ({ value, json }) => ({
+    label: getDefaultLabel({ json }),
+    message: value,
+  }),
+}
+
+const authorMap = {
+  schema: Joi.object({
+    ...commonSchemaFields,
+    user: Joi.object({
+      login: Joi.string().required(),
+    }).required(),
+  }).required(),
+  transform: ({ json }) => json.user.login,
+  render: ({ value }) => ({
+    label: 'author',
+    message: value,
+  }),
+}
+
+const labelMap = {
+  schema: Joi.object({
+    ...commonSchemaFields,
+    labels: Joi.array()
+      .items(
+        Joi.object({
+          name: Joi.string().required(),
+          color: Joi.string().required(),
+        })
+      )
+      .required(),
+  }).required(),
+  transform: ({ json }) => {
+    if (json.labels.length === 0) {
+      throw new InvalidResponse({ prettyMessage: 'no labels found' })
+    }
+    return json.labels.map(l => l.name).join(' | ')
+  },
+  render: ({ value, json }) => {
```

Comment:
What you've written here is fine, though for the purpose of polishing or thinking about ways to structure future work, you could consider this.

It might be a little nicer to have your transform function return an object of props, and your render function to be passed `which` and `props` (and not `json`). That way you can avoid doing rendering in transform and transforming in render. Here your transform function could return e.g. an array of label names and a second array of colors.

Where it really looks odd is the call to `this.render` in the examples, the arguments to which were unexpected.

All totally optional.

Reply:
I now realize that the `labels` badge is doing an odd blending of transform/render 

> It might be a little nicer to have your transform function return an object of props

Not sure whether I'm correctly visualizing what that would look like. I'm thinking that the proposal is for this particular `render` function to return an object (`props`) that's extracted the relevant data portions from the response, and passing that along to `render` (i.e. here the `value` param would be that object instead of the formatted/concatenated label name string). 

The example render felt odd to me too, so let me take another pass at simplifying this.

Reply:
I’m suggesting replacing `value` with an object that contains the values you need.

Reply:
Updated accordingly. Another solid improvement IMO